python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,302 @@
+import copy
+from abc import ABCMeta, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Dict, List, Optional
+
+from google.protobuf.json_format import MessageToDict
+
+from ray.autoscaler._private.resource_demand_scheduler import UtilizationScore
+from ray.autoscaler._private.util import NodeType
+from ray.core.generated.autoscaler_pb2 import ResourceRequest
+from ray.core.generated.instance_manager_pb2 import (
+    NodeTypeConfig,
+    ResourceScheduleConfig,
+    ScheduleResourceBundlesReply,
+    ScheduleResourceBundlesRequest,
+)
+
+
+class SchedulingNodeStatus(Enum):
+    """
+    The status of a scheduling node (`SchedulingNode`)
+    """
+
+    # The node is to be launched.
+    TO_LAUNCH = "TO_LAUNCH"
+    # The node is pending to be launched (a pending request, or a pending instance)
+    PENDING = "PENDING"
+    # The node is running.
+    RUNNING = "RUNNING"
+
+
+@dataclass
+class SchedulingNode:
+    """
+    A abstraction of a node that can be scheduled on by the resource scheduler.
+
+    A scheduling node is expected to be used as:
+
+        node  = SchedulingNode.from_node_config(node_config)
+        infeasible = node.try_schedule(requests)
+        score = node.compute_score()
+
+        .... do something with the score ....
+
+    NOTE:
+        One could also extend the scheduling behavior by overriding:
+            1. try_schedule()
+            2. compute_score()
+
+    """
+
+    # Class level node id counter.
+    _next_node_id = 0
+
+    # A unique id simply for differentiating nodes. Not the same as a normal ray node
+    # id.
+    node_id: int
```

Comment:
why do we need a different id than the ray node id or instance id?

Reply:
To identify a `SchedulerNode` from each other (this is mainly to differentiate nodes from the same node group).